phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/QueryServer.java

```diff
@@ -128,10 +133,40 @@ public QueryServer() {
 
   /** Constructor for use as {@link java.lang.Runnable}. */
   public QueryServer(String[] argv, Configuration conf) {
+    this(argv,conf,null);
+  }
+
+  /** Constructor for use as {@link java.lang.Runnable}. */
+  public QueryServer(String[] argv, Configuration conf,String zookeeperConnectString) {
     this.argv = argv;
     setConf(conf);
+    this.zookeeperConnectString = zookeeperConnectString;
+
+  }
+
+  public void registerToServiceProvider(Configuration conf ) {
+    String basePath = conf.get(QueryServices.PHOENIX_QUERYSERVER_BASE_PATH);
+    String serviceName = conf.get(QueryServices.PHOENIX_QUERYSERVER_SERVICENAME);
+    if (zookeeperConnectString == null) {
+      String zookeeperQuorum = conf.get(QueryServices.ZOOKEEPER_QUORUM_ATTRIB);
+      String zookeeperPort = conf.get(QueryServices.ZOOKEEPER_PORT_ATTRIB);
+      zookeeperConnectString = String.format("%s:%s",zookeeperQuorum,zookeeperPort);
+    }
+    try {
+      registry = new ZookeeperRegistry();
```

Comment:
The QueryServer shouldn't know what concrete implementation its using. I thought I saw you had serviceloader stuff elsewhere to avoid the strong dependency?